### PR TITLE
Fix mask creations of `GPTNeoX` and `GPT2`

### LINKED ISSUE #31943: Very different output depending on whether an attention mask is passed when using caching
### System Info

- `transformers` version: 4.42.0
- Platform: Linux-5.15.0-106-generic-x86_64-with-glibc2.35
- Python version: 3.10.13
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.29.2
- Accelerate config:    not found
- PyTorch version (GPU?): 2.2.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: No
- GPU type: NVIDIA H100 80GB HBM3

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I am debugging issues I experienced when upgrading from `transformers==4.41.2` to `transformers==4.42.x`. Since `4.42.0`, there is a large difference in behavior when passing a cache based on whether you pass an attention mask or not.
```python
import torch
from transformers import AutoModelForSequenceClassification, AutoTokenizer

torch.manual_seed(0)
model_name = "EleutherAI/pythia-14m"
model = AutoModelForSequenceClassification.from_pretrained(model_name, use_cache=True)
tokenizer = AutoTokenizer.from_pretrained(model_name)

cache_inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
cache_outputs = model(**cache_inputs)
print(f"{cache_outputs.logits = }")
# On 4.41.2: tensor([[-1.0874,  1.5391]])
# On 4.42.0: tensor([[-1.0874,  1.5391]])

full_inputs = tokenizer(
    "Hello, my dog is cute. I like to eat ice cream.", return_tensors="pt"
)
len_cache = len(cache_outputs.past_key_values)
noncache_inputs = {
    "input_ids": full_inputs.input_ids[:, len_cache:],
    # "attention_mask": full_inputs.attention_mask,
}
full_outputs = model(**noncache_inputs, past_key_values=cache_outputs.past_key_values)
print(f"{full_outputs.logits = }")
# With attention mask:
# > On 4.41.2: tensor([[-2.9114,  0.3428]])
# > On 4.42.0: tensor([[-2.9107,  0.3432]])
# Without attention mask:
# > On 4.41.2: tensor([[-2.9114,  0.3428]]
# > On 4.42.0: tensor([[-1.7919,  0.4542]]
```

### Expected behavior

I expected the presence of an attention mask not to make a difference when there's no padding, as was the case in `4.41.2`. (I'm also a little surprised that there are small differences even with an attention mask, but they seem less bad currently.)

I'm happy to look into this more but thought I'd post it now in case it was obvious what the issue was.

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?

Fixes #31943

SDPA implementations done by me have introduced a careless mistake in the mask creation, e.g. SDPA has to create a mask even when the attention mask is `None`, for example when cached input is passed as demonstrated in the issue mentioned above.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @amyeroberts

### CODE REVIEW

**Comment by vasqu** [comment]
The timeouts are back :cry:

**Review by ArthurZucker — COMMENTED** [review]
Indeed! Can you add what was reported as an issue to the tests? 🤗

**Comment by vasqu** [comment]
@ArthurZucker I've added the tests to GPTNeoX and GPT2. Not sure if the reference to the issue is a good idea or not, just went with it.

**Review by ArthurZucker — APPROVED** [review]
Thanks for updating, indeed the order makes sense as for full attention you gotta create the mask even if None
